packages/engine/crawlers/solana/transactionHistory.ts

```diff
@@ -0,0 +1,530 @@
+	meta: ParsedTransactionWithMeta['meta'],
+	type: 'sent' | 'received',
+	sender: string,
+	receiver: string,
+	solanaContext: SolanaContext,
+) => {
+	let token: Token = {
+		network: Networks.solana,
+		account: {
+			mint: '',
+			owner: '',
+			address: '',
+			balance: '',
+			decimals: 0,
+		},
+		metadata: {
+			name: 'Unknown',
+			symbol: '',
+			imageUri: '/img/send-token/unknown-token.jpeg',
+		},
+	};
+
+	if (
+		!meta?.preTokenBalances ||
+		!meta?.postTokenBalances ||
+		meta?.postTokenBalances.length === 0 ||
+		meta?.preTokenBalances.length === 0
+	) {
+		const solMetadata = await getMetadata(solanaContext, solMint);
+
+		token = {
+			network: Networks.solana,
+			account: {
+				mint: solMint,
+				owner: '',
+				address: '',
+				balance: '',
+				decimals: 9,
+			},
+			metadata: solMetadata,
+		};
+
+		return token;
+	}
+
+	const tokenBalances = meta.postTokenBalances;
+
+	for (const tokenBalance of tokenBalances) {
+		if (!tokenBalance.owner) continue;
+
+		if (
+			tokenBalance.owner !== FEE_PAYER &&
+			tokenBalance.owner !== (type === 'sent' ? sender : receiver)
+		) {
+			const network = Networks.solana;
+			const account: TokenAccount = {
+				mint: tokenBalance.mint,
+				owner: tokenBalance.owner,
+				address: (instruction as ParsedInstruction).parsed?.info.source,
+				balance: tokenBalance.uiTokenAmount.uiAmount?.toString() || '',
+				decimals: tokenBalance.uiTokenAmount.decimals,
+			};
+			const metadata = await getMetadata(solanaContext, tokenBalance.mint);
+
+			token = {
+				network,
+				account,
+				metadata,
+			};
+
+			return token;
+		}
+	}
+
+	return token;
+};
+
+const getTransactionBalances = (
+	meta: ParsedTransactionWithMeta['meta'],
+	ownerPublicKey: string,
+	type?: 'sent' | 'received',
+) => {
+	if (!meta) {
+		return {
+			preBalance: 0,
+			postBalance: 0,
+		};
+	}
+	if (
+		!meta.preTokenBalances ||
+		!meta.postTokenBalances ||
+		meta.postTokenBalances.length === 0 ||
+		meta.preTokenBalances.length === 0
+	) {
+		if (type === 'sent') {
+			const preBalance = meta.preBalances[0] / 10 ** 9;
+			const postBalance = meta.postBalances[0] / 10 ** 9;
+
+			return {
+				preBalance,
+				postBalance,
+			};
+		}
+
+		return {
+			preBalance: meta.preBalances[1] / 10 ** 9,
+			postBalance: meta.postBalances[1] / 10 ** 9,
+		};
+	} else {
+		let preBalance = 0;
+		let postBalance = 0;
+
+		const preBalances = meta.preTokenBalances.filter(
+			(item) => item.owner === ownerPublicKey,
+		);
+		const postBalances = meta.postTokenBalances.filter(
+			(item) => item.owner === ownerPublicKey,
+		);
+
+		if (!preBalances[0]?.uiTokenAmount.uiAmount) {
+			preBalance = 0;
+		} else {
+			preBalance = preBalances[0].uiTokenAmount.uiAmount;
+		}
+		if (!postBalances[0]?.uiTokenAmount.uiAmount) {
+			postBalance = 0;
+		} else {
+			postBalance = postBalances[0].uiTokenAmount.uiAmount;
+		}
+
+		return {
+			preBalance,
+			postBalance,
+		};
+	}
+};
+
+const getTransactionAmount = (
```

Comment:
`getAmountOfInstruction`?

Reply:
what I see is a function that takes an instruction parameter and calculates based on this parameter, does it make sense?

Reply:
the problem is the definition of the transaction in history, we'll rework it later